Rework Fusion API to use solvers

JehandadKhan opened the pull request:
This PR removes the MetaData Graph logic in MIOpen used to determine the applicable solvers for various fusions. The API and the associated flows remain the same. A few tests have been added to make sure that we are able to ascertain whether solvers work as expected.

JehandadKhan committed:
vector of primitives

JehandadKhan committed:
vector of ops

JehandadKhan committed:
interim commit

averinevg committed:
Non template SolverBase

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Solve merge conflicts, update according to last changes

averinevg committed:
Make SolverBase abstract

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Suppress misleading warnings

averinevg committed:
Fix warning

averinevg committed:
Code decoration

averinevg committed:
Code decoration

averinevg committed:
Suppress misleading warnings for clang only

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Update according to last changes

JehandadKhan committed:
work around for CRTP

JehandadKhan committed:
Merge branch 'develop' into jd/fusion_new_ops

JehandadKhan committed:
Merge branch 'develop' into jd/fusion_new_ops

JehandadKhan committed:
Merge branch 'jd/fusion_new_ops' into jd/non_template_fusion_ops

JehandadKhan committed:
fix merge issues

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

JehandadKhan committed:
Merge branch 'non_template_solver_base' into jd/non_template_fusion_ops

averinevg committed:
use templated mixin, revert some changes

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
fix compiling for pooling solvers

averinevg committed:
suppress -Woverloaded-virtual

averinevg committed:
revert useless changes, add check in find_solution

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

averinevg committed:
Merge branch 'develop' into non_template_solver_base

JehandadKhan committed:
Merge branch 'non_template_solver_base' into jd/non_template_fusion_ops

JehandadKhan committed:
interim tweaks

JehandadKhan committed:
start removing any solver

JehandadKhan committed:
undo last commit, follow solvercontainer model

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan committed:
interim commit

JehandadKhan committed:
first working solver/invoker

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan committed:
Add fusion context

JehandadKhan committed:
Finish second conv solver

JehandadKhan committed:
Add ConvBinWinogradRxSFused solver

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan committed:
split fusion context into context and problem

JehandadKhan committed:
add new fusion solver, add weight attribute to solution

JehandadKhan committed:
Winograd solver flush out

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan committed:
BN + Activ initial solver

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan committed:
fix solver refactor breakages

JehandadKhan committed:
added bn fwd trg solver

JehandadKhan committed:
Add Bwd Bn + Activ solver

JehandadKhan committed:
test updates

JehandadKhan committed:
Add CBA test infra and test asm 1x1 cba solver

JehandadKhan committed:
refactor some test code, add negative nums to test tensors

JehandadKhan committed:
added batchnorm infer test

JehandadKhan committed:
Fix kernel launch data type

JehandadKhan committed:
All tests passing

JehandadKhan committed:
refactor kernel launch args

JehandadKhan committed:
clean up tests

JehandadKhan committed:
clean up, algo logic and clean out op_args

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan committed:
clang format discrepancy

JehandadKhan committed:
add check to fin for Convolution Primitives

JehandadKhan committed:
clang-tidy

JehandadKhan committed:
clang-tidy fixes

JehandadKhan committed:
Fix data type for args

JehandadKhan committed:
more tidy

atamazov commented:
@averinevg Can you please begin reviewing? At top level first; let's understand the general design ideas. Thanks!

JehandadKhan committed:
remove DataCast

daniellowell commented on src/include/miopen/find_solution.hpp:
```diff
@@ -137,7 +137,7 @@ template <class Solver, class Context, class Db>
 FindSolution(Solver s, const Context& context, Db& db, const AnyInvokeParams& invoke_ctx)
 {
-    static_assert(sizeof(Solver) == sizeof(SolverBase), "Solver must be stateless");
+    // static_assert(sizeof(Solver) == sizeof(SolverBase), "Solver must be stateless");
```
Why remove this check?

JehandadKhan replied:
reverted the error

JehandadKhan committed:
restore solver check, fix wino solver args

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan committed:
revert skipping dynamic solutions

averinevg commented on src/include/miopen/fusion.hpp:
```diff
@@ -33,16 +33,22 @@
 #include <vector>
 #include <unordered_map>
 
+#define MIOPEN_CHECK(x)          \
```
@JehandadKhan Is this macro used anywhere? If so, let's move it to some common header.

```suggestion
#define MIOPEN_CHECK_STATUS(x)          \
```

JehandadKhan replied:
This macro is only used in `fusion.cpp` to construct fusion plans to support dedicated CBA API. I have moved it back to the CPP file since its not used anywhere else.

averinevg commented on src/include/miopen/fusion/solvers.hpp:
```diff
@@ -0,0 +1,356 @@
+
+namespace miopen {
+
+struct FusionDescription : SQLiteSerializable<FusionDescription>
```
I suggest moving `FusionDescription` to its own header file inside fusion directory so as not to overload solvers.hpp

atamazov replied:
Yes, just like we do with problem descriptions of other primitives.

JehandadKhan replied:
Done!

averinevg commented on src/include/miopen/fusion/solvers.hpp:
```diff
@@ -0,0 +1,356 @@
+    }
+};
+
+struct FusionContext : miopen::ExecutionContext
```
I suggest moving `FusionContext` to its own header file inside fusion directory so as not to overload solvers.hpp

DrizztDoUrden replied:
I would also suggest changing it from legacy style. For old style API you can just make std::tuple<ExecutionContext, FusionDescription> and use it as a legacy context type.

JehandadKhan replied:
@DrizztDoUrden I am not sure I understand the requirement here. If you would like to remove `FusionDescription` as a member of the `FusionContext` then that would cascade into changes which are beyond the scope of this PR. For example, I would need to rewrite the `FindSolution` function in `find_solution.hpp` which expects the context to have a problem member. Moreover, this design follows the methodology of the `ConvolutionContext`.

Please elaborate as to the target design that you would like to achieve here.

JehandadKhan replied:
`FusionContext` moved to its own file as requested.

DrizztDoUrden replied:
@JehandadKhan What I suggest is the following (like convolution and e.g. activation does it): solvers have methods with "old-style" signatures for temporary compatibility with old `find_solution.hpp` functions which take some type like `std::tuple<ExecutionContext, FusionDescription>` and just map it to methods with correct signatures. Actual implementations have parameters in form of `const ExecutionContext&, const FusionDescription&`. In fact, there are functions which require no mapping at all like this one https://github.com/ROCmSoftwarePlatform/MIOpen/blob/bc74d11db5187873204b623eb2d541d40a501745/src/include/miopen/find_solution.hpp#L227 an you don't need to map anything if you don't use perf db.

JehandadKhan replied:
I get your point, let me rework it.

JehandadKhan replied:
@DrizztDoUrden As discussed, this change would not go with the Find Machinery implemented in `find_solution.hpp` therefore I have implemented in the same style as the tunable convolution solvers. Once the Find machinery is generalized, we can remove the shim between the old style fusion context and "two-parameter" methods. To minimize the changes, I have implemented the two paramter methods in the convolution based fusion solvers.

averinevg commented on src/include/miopen/fusion/solvers.hpp:
```diff
@@ -0,0 +1,356 @@
+
+    /// Tunable solvers provide a GetSolution that takes a Context and PerformanceConfig
+    virtual ConvSolution GetSolution(const FusionContext& ctx, const boost::any& config) const = 0;
+};
+template <class PerformanceConfig>
```
Please add blank line here

JehandadKhan replied:
I dont understand the rationale for that ? I dont believe a blank line here is a standard followed in MIOpen.

averinevg replied:
This is just a recommendation for a better visual experience.

averinevg commented on src/include/miopen/fusion/solvers.hpp:
```diff
@@ -0,0 +1,356 @@
+    bool SetNextValue(const FusionContext& problem);
+    bool IsValid(const FusionContext& problem) const;
+};
+struct ConvBiasActivAsm1x1U : FusionTunableSolver<PerformanceConfigConvBiasActivAsm1x1U>
```
Please add blank line here

averinevg commented on src/include/miopen/fusion_plan.hpp:
```diff
@@ -60,36 +65,23 @@ struct FusionPlanDescriptor : miopenFusionPlanDescriptor
-    bool GetTensorAttr(const std::string& sym, int& val) const;
-
-private:
+    // private:
```
I think it's worth adding a comment or TODO here

JehandadKhan replied:
The commented code has been removed!

averinevg commented on src/include/miopen/problem_description_base.hpp:
```diff
@@ -0,0 +1,35 @@
+ *
+ *******************************************************************************/
+
+#pragma once
```
[Notice] Probably, we should use only either `#include guard` or `#pragma once`. Now we use both, but the first more often.

atamazov replied:
@averinevg This is out of the scope of this PR, I recommend closing this comment or marking it with explicit "[Notice]"

DrizztDoUrden replied:
Still as it's been talked around a lot already, we don't want any new files to have include guards, #pragma once should be used instead.

JehandadKhan replied:
I believe we can mark this comment resolved ?

averinevg commented on src/include/miopen/solver.hpp:
```diff
@@ -38,6 +38,7 @@
 #include <miopen/miopen.h>
 #include <miopen/buffer_info.hpp>
 #include <miopen/performance_config.hpp>
+#include <miopen/batchnorm/problem_description.hpp>
```
solver.hpp header contains base class for solvers and convolution solvers. I think this line should be removed.

JehandadKhan replied:
has been removed.

averinevg commented on src/include/miopen/solver.hpp:
```diff
@@ -364,7 +365,7 @@ struct PerformanceConfigConvAsm1x1U : PerfConfigBase<PerformanceConfigConvAsm1x1
 };
 
-struct ConvAsm1x1U final : ConvTunableSolver<PerformanceConfigConvAsm1x1U>
+struct ConvAsm1x1U : ConvTunableSolver<PerformanceConfigConvAsm1x1U>
```
Why did you remove `final` here?

JehandadKhan replied:
Added back

averinevg commented on src/include/miopen/solver.hpp:
```diff
@@ -420,6 +421,7 @@ struct ConvAsm1x1U final : ConvTunableSolver<PerformanceConfigConvAsm1x1U>
                                         const AnyInvokeParams& invoke_ctx) const;
 };
 
+#if 0
```
If this code is not used, why should we keep it? I think it's worth adding a comment or TODO here

JehandadKhan replied:
Has been removed.

averinevg commented on src/include/miopen/solver_base.hpp:
```diff
@@ -0,0 +1,74 @@
+#pragma once
```
I suggest not to put this in a separate file for now, but to do it later, when the implementation of the new interface for solvers is completed.

JehandadKhan replied:
The file was redundant and has been removed.

averinevg commented on src/ocl/fusionopbiasbnactivocl.cpp:
```diff
@@ -340,6 +166,7 @@ void BatchNormBwdTrainFusionOpDescriptor::calcBNParams(Handle& handle,
 miopenStatus_t BatchNormBwdTrainFusionOpDescriptor::GetNetworkConfig(std::string& network_config,
                                                                      Handle& handle)
 {
+
```
Redundant blank line

JehandadKhan replied:
has been removed.

averinevg reviewed (changes requested):
@JehandadKhan It is difficult to conduct quality review for such large PRs as it's hard to track links between files with lots of changes. Let's limit the number of changed lines to 1000 for non-trivial PRs.

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -261,115 +227,11 @@ miopenStatus_t ConvForwardOpDescriptor::SetArgs(OperatorArgs& args,
-    auto w_any = OpKernelArg(w);
-    args.ins_arg("weights" + std::to_string(GetIdx()), w_any);
-
+    const auto op_args = std::make_shared<fusion::ConvolutionOpInvokeParam>(w);
```
It's possible that I miss something, but should this guys be shared? I mean, are they really copied? Shouldn't they be unique_ptr's? Not that it would matter performance-wise, but ref counting would get eliminated. And you can still pass them around as raw ptrs anyway.

JehandadKhan replied:
I agree, this has been converted to unique_ptrs throughout with a few changes to accomodate the update.

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -810,562 +378,79 @@ miopenStatus_t BiasFusionOpDescriptor::SetArgs(OperatorArgs& args,
-        auto op   = op_map.at(idx);
-        auto keys = op->GetArgs();
-        for(auto&& key_arg : keys)
+    const auto solvers    = solver::SolverContainer<solver::fusion::ConvBiasActivAsm1x1U,
```
You should move the SolverContainer definition to a separate function and just call it here. You can mark it as inline if you want to although it would probably get inlined anyway.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -810,562 +378,79 @@ miopenStatus_t BiasFusionOpDescriptor::SetArgs(OperatorArgs& args,
-        {
-            MIOPEN_THROW("Kernel arguments not setup properly");
+            const auto strAlgo =
+                miopen::solver::Id{sol.solver_id}.GetAlgo(miopen::conv::Direction::Forward);
```
Save solver::Id{...} to a local.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -810,562 +378,79 @@ miopenStatus_t BiasFusionOpDescriptor::SetArgs(OperatorArgs& args,
+    else
+        sols = tmp_sols;
+    auto net_config =
+        input_desc.ToString() + ((input_desc.GetType() == miopenHalf) ? "FP16" : "FP32");
```
Generating a network config should be done in a separate method and via a string stream rather than concatenation. Also it should probably happen after the "UsupportedOp" check, but I may be wrong here.

JehandadKhan replied:
Agreed on both, I have updated the code.

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -810,562 +378,79 @@ miopenStatus_t BiasFusionOpDescriptor::SetArgs(OperatorArgs& args,
+                MIOPEN_THROW(miopenStatusInternalError,
+                             "Invoker missing from solver " + sol.solver_id);
+            const auto invoker = handle.PrepareInvoker(
+                *sol.invoker_factory, sol.construction_params); // force compilation
```
Note: this doesn't force a compilation, kernels may be read from the binary cache or from HIP/OCL program handle cache in KCache object in the handle. This only forces the creation an invoker. Shouldn't you check if one already exists with your netcfg?

JehandadKhan replied:
I am not sure I fully understand. The call to `PrepareInvoker` includes a call to `AddKernel` which can force the creation of a `Program` object. Therefore, I was under the impression that the `AddKernel` logic you are referring to is not necessary. Can you please elaborate what am I missing here ?

DrizztDoUrden replied:
`AddKernel` does not necessarily compile anything: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/bc74d11db5187873204b623eb2d541d40a501745/src/kernel_cache.cpp#L130
So this may call compilation, but it may as well not and does not "force" it.
And by checking if anything is pre-existing I was talking of adding something like this:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/bc74d11db5187873204b623eb2d541d40a501745/src/include/miopen/find_solution.hpp#L350
at the start of the method to prevent duplicate find and re-creation of invokers if compile has been called twice unless that is intended and wished behaviour.

atamazov replied:
@DrizztDoUrden @JehandadKhan 

👍 Yeah, we need to check if Invoker already exists before creating it.

Some clarifications. This is true:
> The call to PrepareInvoker includes a call to AddKernel... 

This is not:
> which can force the creation of a Program object

Creation of a Program object happens only if the same Program object does not exist. If it exists, it's simply reused.

Otherwise the library creates it by calling `handle.LoadProgram(...)`. The latter method may reuse a code object from the binary cache. 

That is, compilation happens only if:
- Program object does not exist (not registered in the program_map)
- Code object does not reside in the binary cache.

I recommend removing the `// force compilation` comment.

JehandadKhan replied:
Removed

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -1388,51 +468,21 @@ miopenStatus_t FusionPlanDescriptor::Execute(const Handle& handle,
-    kernel(args);
+    const auto invoker = handle.GetInvoker(network_config, solver::Id{solution.solver_id}, {});
+    const auto& tmp    = op_args.params[1];
+    const auto& tmp2   = std::dynamic_pointer_cast<fusion::ActivationOpInvokeParam>(tmp);
```
Do you need the cast here or you can just pass AnyInvokeParams around? They were designed to be passed around like that and only have specific type on creation and inside of the invoker.
Plus you can do `dynamic_cast<...>(*tmp)`. That would eliminate ref count updates.

JehandadKhan replied:
due to the unique_ptr update, this has been fixed.

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -1388,51 +468,21 @@ miopenStatus_t FusionPlanDescriptor::Execute(const Handle& handle,
+    const auto invoker = handle.GetInvoker(network_config, solver::Id{solution.solver_id}, {});
+    const auto& tmp    = op_args.params[1];
+    const auto& tmp2   = std::dynamic_pointer_cast<fusion::ActivationOpInvokeParam>(tmp);
+    const auto tmp3    = op_args.params[1];
```
Are all this values not used at all? Is there any reason to create one set of "refs" (tmp2 is not really a ref as it owns the data it gets initialized from) and one of copies?

JehandadKhan replied:
This was redundant code and has been removed

DrizztDoUrden commented on src/include/miopen/conv/problem_description.hpp:
```diff
@@ -142,9 +143,10 @@ constexpr TElement GetW5(int spatial_dims, const std::vector<TElement>& data)
 namespace conv {
 
-struct ProblemDescription
+struct ProblemDescription : public ProblemDescriptionBase
```
Please, use public everywhere either explicitly or implicitly.

JehandadKhan replied:
removed `public`

DrizztDoUrden commented on src/include/miopen/fusion.hpp:
```diff
@@ -52,13 +58,15 @@ enum FusionKernelSourceType
     friend std::ostream& operator<<(std::ostream& stream, const OperatorArgs& x);
-    std::vector<OpKernelArg> args_vec;
-    std::unordered_map<std::string, OpKernelArg> args_map;
+    void SetArg(size_t idx, const std::shared_ptr<fusion::FusionOpInvokeParamBase>& arg)
+    {
+        if(params.size() < (idx + 1))
+            params.resize(idx + 1);
+        params[idx] = arg;
+    }
```
```suggestion
    void SetArg(size_t idx, std::shared_ptr<fusion::FusionOpInvokeParamBase> arg)
    {
        if(params.size() < (idx + 1))
            params.resize(idx + 1);
        params[idx] = std::move(arg);
    }
```
This allows both move and copy. Also one may want to enforce move here with &&. And I'd rather use unique_ptr here anyway.

JehandadKhan replied:
Changed to unique_ptr

DrizztDoUrden commented on src/include/miopen/fusion/fusion_invoke_params.hpp:
```diff
@@ -0,0 +1,187 @@
+                       TensorDescriptor out_desc,
+                       Data_t out_,
+                       bool gfx90aFp16alt_)
+        // : op_invokers(std::move(op_invokers_)),
```
This should be move. You copy it two times in a row.

JehandadKhan replied:
removed since incompatible with `unique_ptr`

DrizztDoUrden commented on src/include/miopen/fusion/fusion_invoke_params.hpp:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    FusionInvokeParams(InvokeType type_,
+                       std::vector<std::shared_ptr<FusionOpInvokeParamBase>> op_invokers_,
```
This should probably be unique_ptr's

JehandadKhan replied:
Done

DrizztDoUrden commented on src/include/miopen/fusion/solvers.hpp:
```diff
@@ -0,0 +1,356 @@
+        const auto conv_prob = problem.GetConvProblem(idx, dir);
+        if(dir == conv::Direction::Forward)
+        {
+            auto ctx                       = ConvolutionContext{conv_prob};
```
Field-by-field copy is not a good approach. This may lead to very hard to find errors in the future, e.g. if some crucial field would be added to the original type.

JehandadKhan replied:
Fixed.

JehandadKhan replied:
> I would also suggest changing it from legacy style. For old style API you can just make std::tuple<ExecutionContext, FusionDescription> and use it as a legacy context type.

@DrizztDoUrden  The issue with trying to do that is then I cant use `FindSolution` in the `find_solution.hpp` which requires `problem` to be a member of the Context. Since `ConvolutionContext` follows the same design as `FusionContext` I am not sure why I would want to deviate unless we would like to change both, which is best addressed in a separate PR.

Please elaborate if I have missed the point.

DrizztDoUrden commented on src/include/miopen/fusion/utils.hpp:
```diff
@@ -0,0 +1,126 @@
+            return idx;
+        ++idx;
+    }
+    return -1;
```
```suggestion
    auto it = std::find_if(op_map.cbegin(), op_map.cend(), [op](auto&& item){ return item->kind() == op; });
    return it == op_map.cend() ? -1 : std::distance(op_map.cbegin(), it);
```
Or if you crave for cpu ticks, you should use a regular for loop instead, like `for (auto idx = 0; idx < op_map.size(); ++idx)`. Maybe even caching op_map.size(). But std::find_if should not differ by a measurable amount inside the library.

JehandadKhan replied:
Fixed.

DrizztDoUrden commented on src/include/miopen/fusion/utils.hpp:
```diff
@@ -0,0 +1,126 @@
+                return idx;
+            ++idx;
+        }
+        return -1;
```
std::find_if and std::distance

JehandadKhan replied:
Fixed

DrizztDoUrden commented on src/solver.cpp:
```diff
@@ -192,8 +194,11 @@ miopenConvAlgorithm_t Id::GetAlgo() const
+                     uint64_t value,
+                     Primitive primitive,
+                     const std::string& str,
+                     miopenConvAlgorithm_t algo = miopenConvolutionAlgoDirect)
```
This is also used by solvers, which are not convolution. It should not set convAlgo for them.

JehandadKhan replied:
Fixed.

DrizztDoUrden commented on src/solver/batchnorm/backward_per_activation.cpp:
```diff
@@ -32,11 +32,208 @@
 
+namespace fusion {
+
+bool BnBwdTrgActivationFused::IsApplicable(const FusionContext& context) const
```
I would suggest moving fused solver to a separate file.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/solver/batchnorm/forward_inference.cpp:
```diff
@@ -32,11 +32,146 @@
 
+namespace fusion {
+
+bool BnFwdInferActivationFused::IsApplicable(const FusionContext& context) const
```
I would suggest moving fused solver to a separate file.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/solver/batchnorm/forward_per_activation.cpp:
```diff
@@ -32,10 +32,216 @@
 namespace solver {
+namespace fusion {
+
+bool BnFwdTrgActivationFused::IsApplicable(const FusionContext& context) const
```
I would suggest moving fused solver to a separate file.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/solver/conv_bin_winoRxS.cpp:
```diff
@@ -497,25 +502,64 @@ ConvSolution ConvBinWinogradRxS::GetSolution(const ExecutionContext& ctx,
 
-bool ConvBinWinogradRxSFused::IsApplicable(const ExecutionContext& ctx,
-                                           const ProblemDescription& problem) const
+bool ConvBinWinogradRxSFused::IsApplicable(const FusionContext& params) const
```
I would suggest moving fused solver to a separate file.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/solver/conv_ocl_dir2Dfwd.cpp:
```diff
@@ -484,12 +489,191 @@ ConvSolution ConvOclDirectFwd::GetSolution(const ConvolutionContext& ctx,
-ConvSolution ConvOclDirectFwdFused::GetSolution(const ConvolutionContext& ctx,
-                                                const ProblemDescription& problem,
-                                                const LegacyPerformanceConfig& config) const
+namespace fusion {
```
I would suggest moving fused solver to a separate file.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/solver/conv_ocl_dir2Dfwd.cpp:
```diff
@@ -484,12 +489,191 @@ ConvSolution ConvOclDirectFwd::GetSolution(const ConvolutionContext& ctx,
+    const int bn_idx    = GetOpIdx(desc.op_map, miopenFusionOpBatchNormInference);
+
+    if(bias_idx != -1)
+        kernel_info.comp_options += " -DMLO_CONV_BIAS=" + std::to_string(1);
```
Please use `KernelBuildParams` in all new code.

JehandadKhan replied:
Done

DrizztDoUrden commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -948,19 +950,43 @@ ConvSolution ConvBinWinogradRxSf2x3g1::GetSolution(const ConvolutionContext& ctx
 }
 
-bool ConvBinWinogradRxSf2x3g1Fused::IsApplicable(const ConvolutionContext&) const
+namespace fusion {
```
I would suggest moving fused solver to a separate file.

JehandadKhan replied:
Done

JehandadKhan committed:
refactor code in light of PR comments

JehandadKhan commented on .githooks/pre-commit:
```diff
@@ -4,7 +4,7 @@
 # the staged changes.
 
-base=clang-format-10
+base=clang-format-12
```
[Note]: This is leftover from previous commit and may be moved to a separate PR if needed.

JehandadKhan committed:
move solver to separate file, fix Copyright year

JehandadKhan committed:
Merge branch 'develop' into jd/non_template_fusion_ops

JehandadKhan commented:
@DrizztDoUrden , @averinevg  Thanks for your review. I have addressed all the comments, please take a look so we can push this PR through. 

cc @junliume

JehandadKhan committed:
update tests to the new object structures

averinevg commented on src/solver/conv_bin_winoRxS_fused.cpp:
```diff
@@ -0,0 +1,240 @@
+ *
+ *******************************************************************************/
+
+#include <miopen/solver.hpp>
```
@JehandadKhan This `#include` directive can be removed

JehandadKhan replied:
Done

averinevg commented on src/solver/conv_bin_winoRxS.cpp:
```diff
@@ -32,6 +32,9 @@
 #include <miopen/kernel_build_params.hpp>
 #include <miopen/conv/data_invoke_params.hpp>
 #include <miopen/conv/tensors.hpp>
+#include <miopen/fusion_plan.hpp>
+#include <miopen/fusion/solvers.hpp>
+#include <miopen/fusion/utils.hpp>
```
Redundant ~~headers~~ `#include` directives

atamazov replied:
@averinevg I guess you mean not _headers_ but `#include` directives? 😉

averinevg replied:
@atamazov Yes, you are right

JehandadKhan replied:
Removed

averinevg commented on src/solver/conv_bin_winoRxS.cpp:
```diff
@@ -42,6 +45,8 @@ MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_AMD_WINOGRAD_RXS_WRW)
 /// \todo Detect at runtime and remove this var:
 MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_SRAM_EDC_DISABLED)
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_AMD_FUSED_WINOGRAD)
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_GCN_ASM_KERNELS)
```
Redundant env variables

JehandadKhan replied:
Removed

averinevg commented on src/solver/conv_asm_1x1u_bias_activ.cpp:
```diff
@@ -35,170 +35,243 @@
 #include <miopen/generic_search.hpp>
```
@JehandadKhan I suggest to rename source file to `conv_asm_1x1u_bias_activ_fused.cpp` using `git mv`

atamazov replied:
@averinevg This PR does not add this file, only modifies it. Therefore your recommendation is good but out of the scope. Let's not add extra work to this big PR ;)

averinevg replied:
@atamazov This PR also splits fused and non-fused solvers with adding some new files with `_fused` ending. Renaming one file is not much work compared to the rest.

atamazov replied:
@averinevg Ok, let's hope that github will not show this as one removed file plus one added file (which happens from time to time when you rename and edit a file in the same PR)))

averinevg replied:
@atamazov This is what `git mv` is for :)

JehandadKhan replied:
Renamed

averinevg replied:
@atamazov You were right, github shows this as renaming in a single commit and as 2 files for the whole PR

averinevg commented on src/solver/conv_ocl_dir2Dfwd_fused.cpp:
```diff
@@ -0,0 +1,233 @@
+
+#include <miopen/handle.hpp>
+#include <miopen/legacy_exhaustive_search.hpp>
+#include <miopen/solver.hpp>
```
@JehandadKhan This `#include` directive can be removed

JehandadKhan replied:
Removed

averinevg commented on src/solver/conv_ocl_dir2Dfwd.cpp:
```diff
@@ -267,9 +267,9 @@ bool ConvOclDirectFwd::IsValidPerformanceConfig(const ProblemDescription& proble
-                                    const LegacyPerformanceConfig& config)
+ConvSolution BaseGetSolution(const ConvolutionContext& ctx,
+                             const ProblemDescription& problem,
+                             const LegacyPerformanceConfig& config)
```
@JehandadKhan This function related to `ConvOclDirectFwd` solver, I suggest to rename it or made a static member of the class.

atamazov replied:
Yeah, BaseGetSolution is too generic. Let's make it a static member of one of these classes (`ConvOclDirectFwd` or `ConvOclDirectFwdFused`). Or prepend its name with `ConvOclDirectFwd`.

JehandadKhan replied:
Converted to a static method

averinevg commented on src/include/miopen/fusion/solvers.hpp:
```diff
@@ -0,0 +1,206 @@
+    const std::string& SolverDbId() const override { return GetSolverDbId<ConvBiasActivAsm1x1U>(); }
+
+    bool IsApplicable(const FusionContext& problem) const override;
+    ConvSolution
```
@JehandadKhan Probably `GetWorkspaceSize()` and `MayNeedWorkspace()` are missing

atamazov replied:
@averinevg If WS is not needed, then the solver can safely derive them from `SolverMixin<FusionContext>`, right?

averinevg replied:
@atamazov Yes, but this solver comes from `src/include/miopen/solver.hpp`, and these functions are there

JehandadKhan replied:
The Fusion solvers are derived from `SolverMixin<FusionContext>` and inherit both of these methods since it derives from `SolverBase`

averinevg replied:
@JehandadKhan `ConvBiasActivAsm1x1U` uses `GetSolution` from `ConvAsm1x1U`, and `ConvAsm1x1U` has these methods

DrizztDoUrden commented on src/fusion.cpp:
```diff
@@ -810,562 +380,91 @@ miopenStatus_t BiasFusionOpDescriptor::SetArgs(OperatorArgs& args,
-            MIOPEN_THROW("Kernel arguments not setup properly");
+            const auto id      = miopen::solver::Id{sol.solver_id};
+            const auto strAlgo = id.GetAlgo(miopen::conv::Direction::Forward);
+            MIOPEN_LOG_I2(miopen::solver::Id{sol.solver_id}.ToString());
```
Duplicate constructor

JehandadKhan replied:
Fixed

DrizztDoUrden commented on src/include/miopen/fusion/context.hpp:
```diff
@@ -0,0 +1,92 @@
+    {
+    }
+    void GetNetworkConfig(std::string& net_config) const
+    {
```
I would suggest using a string stream here.

JehandadKhan replied:
Updated the `GetNetworkConfig` to use `std::stringstream` everywhere

JehandadKhan committed:
fix tests with new internal

JehandadKhan committed:
second round of reviews

JehandadKhan committed:
interim commit

JehandadKhan committed:
revert some changes to address find not working with separate context and problem description

JehandadKhan committed:
fix tidy issues

JehandadKhan committed:
more tidy checks

JehandadKhan commented:
@DrizztDoUrden @averinevg @atamazov Please let me know if there are any other issues.

DrizztDoUrden commented on src/include/miopen/any_solver.hpp:
```diff
@@ -27,6 +27,7 @@
 #ifndef MIOPEN_GUARD_MLOPEN_ANY_SOLVER_HPP
 #define MIOPEN_GUARD_MLOPEN_ANY_SOLVER_HPP
 
+#include <miopen/problem_description_base.hpp>
```
Not critical, but is this used in the header?

DrizztDoUrden reviewed (approved):
LGTM (my comments resolved)
